builtin/providers/azurerm/resource_arm_app_gateway.go

```diff
@@ -0,0 +1,1667 @@
+			},
+		}
+
+		if host := data["host_name"].(string); host != "" {
+			listener.ApplicationGatewayHTTPListenerPropertiesFormat.HostName = &host
+		}
+
+		if sslCertName := data["ssl_certificate_name"].(string); sslCertName != "" {
+			certID := fmt.Sprintf("%s/sslCertificates/%s", gatewayID, sslCertName)
+			listener.ApplicationGatewayHTTPListenerPropertiesFormat.SslCertificate = &network.SubResource{
+				ID: &certID,
+			}
+		}
+
+		if requireSNI, ok := data["require_sni"].(bool); ok {
+			listener.ApplicationGatewayHTTPListenerPropertiesFormat.RequireServerNameIndication = &requireSNI
+		}
+
+		httpListeners = append(httpListeners, listener)
+	}
+
+	return &httpListeners
+}
+
+func expandAppGatewayProbes(d *schema.ResourceData) *[]network.ApplicationGatewayProbe {
+	configs := d.Get("probe").([]interface{})
+	backendSettings := make([]network.ApplicationGatewayProbe, 0, len(configs))
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		name := data["name"].(string)
+		protocol := data["protocol"].(string)
+		probePath := data["path"].(string)
+		host := data["host"].(string)
+		interval := int32(data["interval"].(int))
+		timeout := int32(data["timeout"].(int))
+		unhealthyThreshold := int32(data["unhealthy_threshold"].(int))
+
+		setting := network.ApplicationGatewayProbe{
+			Name: &name,
+			ApplicationGatewayProbePropertiesFormat: &network.ApplicationGatewayProbePropertiesFormat{
+				Protocol:           network.ApplicationGatewayProtocol(protocol),
+				Path:               &probePath,
+				Host:               &host,
+				Interval:           &interval,
+				Timeout:            &timeout,
+				UnhealthyThreshold: &unhealthyThreshold,
+			},
+		}
+
+		backendSettings = append(backendSettings, setting)
+	}
+
+	return &backendSettings
+}
+
+func expandAppGatewayRequestRoutingRules(d *schema.ResourceData, gatewayID string) *[]network.ApplicationGatewayRequestRoutingRule {
+	configs := d.Get("request_routing_rule").([]interface{})
+	rules := make([]network.ApplicationGatewayRequestRoutingRule, 0, len(configs))
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		name := data["name"].(string)
+		ruleType := data["rule_type"].(string)
+		httpListenerName := data["http_listener_name"].(string)
+		httpListenerID := fmt.Sprintf("%s/httpListeners/%s", gatewayID, httpListenerName)
+
+		rule := network.ApplicationGatewayRequestRoutingRule{
+			Name: &name,
+			ApplicationGatewayRequestRoutingRulePropertiesFormat: &network.ApplicationGatewayRequestRoutingRulePropertiesFormat{
+				RuleType: network.ApplicationGatewayRequestRoutingRuleType(ruleType),
+				HTTPListener: &network.SubResource{
+					ID: &httpListenerID,
+				},
+			},
+		}
+
+		if backendAddressPoolName := data["backend_address_pool_name"].(string); backendAddressPoolName != "" {
+			backendAddressPoolID := fmt.Sprintf("%s/backendAddressPools/%s", gatewayID, backendAddressPoolName)
```

Comment:
given there's no guarantee that these URL's won't change in the future, would it make more sense to pass in the Backend Address Pool ID, rather than the name?